core/src/main/java/org/elasticsearch/common/xcontent/ObjectParser.java

```diff
@@ -214,6 +220,58 @@ public void declareField(Parser<Value, Context> p, ParseField parseField, ValueT
         declareField((p, v, c) -> consumer.accept(v, parser.parse(p, c)), parseField, type);
     }
 
+    /**
+     * Declares a parser for fields of which names are not known upfront when creating the ObjectParser.
+     * As an example, in the aggregation output we have things like:
+     * <pre><code>
+     * "aggregations" : {
+     *   "genres" : {  <--- arbitrary aggregation name
+     *      "doc_count_error_upper_bound": 0,
+     *      "sum_other_doc_count": 0,
+     *      "buckets" : [
+     *          {
+     *              "key" : "jazz",
+     *              "doc_count" : 10
+     *              "total_number_of_ratings": { <--- arbitrary aggregation name
+     *                  "value": 2691
+     *              }
+     *          },
+     *          [...]
+     *      ]
+     *   }
+     * }
+     *
+     * Since field names are arbitrary in these cases, we cannot match them with a ParseField like we do
+     * in other places when using ObjectParser. This method can be used to register a "fall-back"
+     * parser for any field that isn't matched by any previous field parser.
+     * It accepts a dedicated ContextParser and a consumer for the value that is produced.
+     *
+     * @param consumer handle the values once they have been parsed
+     * @param parser parses each nested object
+     * @param type the accepted values for this field
+     */
+    public <T> void declareUnknownFieldParser(BiConsumer<Value, T> consumer, ContextParser<Context, T> parser, ValueType type) {
+        Objects.requireNonNull(consumer, "[consumer] is required");
+        Objects.requireNonNull(parser, "[parser] is required");
+        Objects.requireNonNull(type, "[type] is required");
+        this.unknownFieldParser = new FieldParser((p, v, c) -> consumer.accept(v, parser.parse(p, c)), type.supportedTokens(),
+                new MatchAllParseField(), type);
+    }
+
+    /**
+     * A special ParseField that matches all field names
+     */
+    private class MatchAllParseField extends ParseField {
```

Comment:
I'm not super comfortable making this. I think maybe instead we should add the match skipping at the `FieldParser` level. Maybe some kind of subclass that skips or something. Not sure.